### PR TITLE
[ML][Data Frame] responding with 409 status code when failing _stop

### LINKED ISSUE #44103: [ML] Incorrect response when trying to stop a failed data frame transform without force=true
Spotted in 7.3.0

When trying to stop a failed data frame transform without specifying `force=true` in query params, the status code and format of the response is not what you'd expect.

Currently the API responds with **200** and the body looks like this:
```json
 {
  "task_failures" : [ {
    "task_id" : 15302,
    "node_id" : "1TYVPNuERNe40f9tTRPVAQ",
    "status" : "CONFLICT",
    "reason" : {
      "type" : "status_exception",
      "reason" : "Unable to stop data frame transform [some_data_frame_id] as it is in a failed state with reason: [task encountered more than 10 failures; latest failure: Failed to retrieve checkpoint]. Use force stop to stop the data frame transform."
    }
  } ],
  "acknowledged" : false
} 
```

In order to be consistent with the rest of the ML APIs, what we should be receiving instead should be a **409** and a body that looks like this:
```json
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Unable to stop data frame transform [some_data_frame_id] as it is in a failed state. Use force stop to stop the data frame transform."
            }
        ],
        "type": "status_exception",
        "reason": "Unable to stop data frame transform [some_data_frame_id] as it is in a failed state. Use force stop to stop the data frame transform."
    },
    "status": 409
}
```

**Steps to reproduce**
1. Create a data frame
2. Start it and cause it to fail. (deleting the source index is one way to do it)
3. After the task state becomes `failed`, send the `_stop` request without specifying `force=true` as a query parameter.
4. Notice that the response status code and body format is not consistent with the usual error response format in ES APIs.

### PR DESCRIPTION
**Pull request by benwtrent**
When attempting a `_stop` call against any number of tasks with `force=false` and any of the tasks are failed, we should not stop any of the tasks. The response code should indicate an error with details in the body.

closes #44103

### CODE REVIEW

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameTaskFailedStateIT.java [hunk 1]
```diff
@@ -30,53 +29,47 @@ public void testFailureStateInteraction() throws Exception {
         createReviewsIndex();
         String transformId = "failure_pivot_1";
         String dataFrameIndex = "failure_pivot_reviews";
-        createPivotReviewsTransform(transformId, dataFrameIndex, null);
-        deleteIndex(REVIEWS_INDEX_NAME); // trigger start failure due to index missing
+        createContinuousPivotReviewsTransform(transformId, dataFrameIndex, null);
         startDataframeTransform(transformId, false);
+        // wait for our initial indexing to complete
```
Comment (benwtrent) [inline x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameTaskFailedStateIT.java]: These test changes are in here as I was able to get the test to pass with other code changes. However, those other changes may not need to exist within 7.3. Keeping the test changes for future help when we enable this test again but keeping it muted for now.

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java [hunk 2]
```diff
@@ -88,8 +91,31 @@ protected void doExecute(Task task, StopDataFrameTransformAction.Request request
...
+                            PersistentTasksCustomMetaData.PersistentTask<?> dfTask = tasks.getTask(transformId);
+                            if (dfTask.getState() instanceof DataFrameTransformState
+                                && ((DataFrameTransformState) dfTask.getState()).getTaskState() == DataFrameTransformTaskState.FAILED) {
+                                failedTasks.add(transformId);
+                                failedReasons.add(((DataFrameTransformState) dfTask.getState()).getReason());
+                            }
+                        }
+                        if (failedTasks.isEmpty() == false) {
+                            String msg = failedTasks.size() == 1 ?
+                                "Unable to stop data frame transform [" + request.getId()
```
Comment (davidkyle) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java]: nit the same text is used in `taskOperation()`, could be extracted and refactored.

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameTaskFailedStateIT.java (hunk 1 above)
Comment (davidkyle) [reply x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameTaskFailedStateIT.java]: I don't think this test is necessary. It goes to great lengths to set up a transform that can fail when all we really want to do is test that if the cluster state has failed data frame persistent tasks an error and 409 is returned.

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java [hunk 3]
```diff
@@ -108,6 +134,8 @@ protected void taskOperation(StopDataFrameTransformAction.Request request, DataF
         }
 
         if (ids.contains(transformTask.getTransformId())) {
+            // This should not occur as we validate that none of the tasks are in a failed state earlier
+            // Keep this check in here for insurance.
```
Comment (davidkyle) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java]: 👍

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java [hunk 4]
```diff
@@ -88,8 +91,31 @@ protected void doExecute(Task task, StopDataFrameTransformAction.Request request
                 new PageParams(0, 10_000),
                 request.isAllowNoMatch(),
                 ActionListener.wrap(hitsAndIds -> {
+                    if (request.isForce() == false) {
+                        PersistentTasksCustomMetaData tasks = state.metaData().custom(PersistentTasksCustomMetaData.TYPE);
```
Comment (davidkyle) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java]: This logic is the main change. If you really want to test this code extract it to a static method and it is easily unit testable. `MlTasksTests` is a good reference for building `PersistentTasksCustomMetaData`

**Review by davidkyle — COMMENTED** [review]
Looks good. Left a few comments I don't think you need the integration test this is mergeable without it

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java [hunk 5]
```diff
@@ -88,8 +91,31 @@ protected void doExecute(Task task, StopDataFrameTransformAction.Request request
...
+                        List<String> failedTasks = new ArrayList<>();
+                        List<String> failedReasons = new ArrayList<>();
+                        for (String transformId : hitsAndIds.v2()) {
+                            PersistentTasksCustomMetaData.PersistentTask<?> dfTask = tasks.getTask(transformId);
+                            if (dfTask.getState() instanceof DataFrameTransformState
```
Comment (davidkyle) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java]: You need a `null` check here. That's why the tests failed

**Comment by benwtrent** [comment]
run elasticsearch-ci/bwc
run elasticsearch-ci/default-distro

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameRestTestCase.java [hunk 6]
```diff
@@ -161,7 +161,8 @@ protected void createContinuousPivotReviewsTransform(String transformId, String
 
         String config = "{ \"dest\": {\"index\":\"" + dataFrameIndex + "\"},"
             + " \"source\": {\"index\":\"" + REVIEWS_INDEX_NAME + "\"},"
-            + " \"sync\": {\"time\":{\"field\": \"timestamp\", \"delay\": \"15m\"}},"
+            //Set frequency high for testing
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameRestTestCase.java]: nit: "Set frequency high" - DYM low? The default is 60s, so 1s is lower?

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java [hunk 7]
```diff
@@ -63,6 +69,32 @@ public TransportStopDataFrameTransformAction(TransportService transportService,
...
+                    failedReasons.add(((DataFrameTransformState) dfTask.getState()).getReason());
+                }
+            }
+            if (failedTasks.isEmpty() == false) {
+                String msg = failedTasks.size() == 1 ?
+                    DataFrameMessages.getMessage(DATA_FRAME_CANNOT_STOP_FAILED_TRANSFORM,
+                        failedTasks.get(0),
+                        failedReasons.get(0)) :
+                    "Unable to stop data frame transforms. The following transforms are in a failed state " +
+                        failedTasks + " with reasons " + failedReasons + ". Use force stop to stop the data frame transforms.";
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java]: nit: use a constant

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java [hunk 8]
```diff
@@ -88,8 +120,9 @@ protected void doExecute(Task task, StopDataFrameTransformAction.Request request
                 new PageParams(0, 10_000),
                 request.isAllowNoMatch(),
                 ActionListener.wrap(hitsAndIds -> {
+                    validateTaskState(state, hitsAndIds.v2(), request.isForce());
```
Comment (hendrikmuhs) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportStopDataFrameTransformAction.java]: nit: taskstate -> taskstates (or is it tasksstates?)

**Review by hendrikmuhs — APPROVED** [review]
LGTM, some very minor comments

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameRestTestCase.java (hunk 6 above)
Comment (droberts195) [reply x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameRestTestCase.java]: High frequency => low time between repeats

(Similar to high frequency => short wavelength)

But if it's confusing then maybe bypass this by changing the comment to something like `// Set frequency to 1s to make the test run as fast as possible`.

Comment (davidkyle) [reply x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameRestTestCase.java]: This is one of those pedantic things that annoy me; frequency is measured in Hz not seconds, this is an interval. However, most people will read this and understand that frequency represents the period here and is probably the least confusing naming option.

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameRestTestCase.java (hunk 6 above)
Comment (benwtrent) [reply x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFrameRestTestCase.java]: A high frequency in Hz means a short distance between cycles. 

A high frequency here means a short time between attempts.

**Pull request merged**